File: openFaces/source/org/openfaces/renderkit/timetable/daySwitcher.js
/*
 * OpenFaces - JSF Component Library 2.0
 * Copyright (C) 2007-2010, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */

O$.DaySwitcher = {
  _init: function(daySwitcherId,
                  dayTableId,
                  day,
                  pattern,
                  upperPattern,
                  locale,
                  stylingParams,
                  enabled) {
    var dtf = O$.getDateTimeFormatObject(locale);
    var daySwitcher = O$.initComponent(daySwitcherId, {rollover: stylingParams.rolloverClass}, {
      _day: dtf.parse(day, "dd/MM/yyyy"),
      _dayTableId: dayTableId,
      _pattern: pattern,
      _upperPattern: upperPattern,
      _locale: locale
    });

    var textId = daySwitcherId + "::text";
    daySwitcher._text = O$(textId);
    if (daySwitcher._text) {
      O$.initComponent(textId, {rollover: stylingParams.textRolloverClass});
    }

    var upperTextId = daySwitcherId + "::upper_text";
    daySwitcher._upperText = O$(upperTextId);
    if (daySwitcher._upperText) {
      O$.initComponent(upperTextId, {rollover: stylingParams.upperTextRolloverClass});
    }

    if (enabled) {
      var previousButtonId = daySwitcherId + "::previous_button";
      O$.initComponent(previousButtonId, {rollover: stylingParams.previousButtonRolloverClass});
      daySwitcher._previousButton = O$(previousButtonId);
      var previousButton = daySwitcher._previousButton;

      var nextButtonId = daySwitcherId + "::next_button";
      O$.initComponent(nextButtonId, {rollover: stylingParams.nextButtonRolloverClass});
      daySwitcher._nextButton = O$(nextButtonId);
      var nextButton = daySwitcher._nextButton;

      previousButton.onmouseup = function () {
        O$.setStyleMappings(previousButton, {
          pressed: null
        });
      };

      nextButton.onmouseup = function () {
        O$.setStyleMappings(nextButton, {
          pressed: null
        });
      };

      previousButton.ondragstart =
      nextButton.ondragstart =
      previousButton.onselectstart =
      nextButton.onselectstart = function(e) {
        O$.breakEvent(e);
      };

      previousButton.ondblclick =
      nextButton.ondblclick = O$.repeatClickOnDblclick;

    }

    daySwitcher._getDayTable = function() {
      if (!daySwitcher._dayTable) {
        daySwitcher._dayTable = O$(daySwitcher._dayTableId);
      }
      return daySwitcher._dayTable;
    };

    daySwitcher.getDay = function() {
      return daySwitcher._day;
    };


    daySwitcher.setDay = function(day) {
      if (O$._datesEqual(daySwitcher._day, day))
        return;
      daySwitcher._day = day;

      if (daySwitcher._pattern) {
        var dtf = O$.getDateTimeFormatObject(daySwitcher._locale);
        daySwitcher._text.innerHTML = dtf.format(day, daySwitcher._pattern);
      }

      if (daySwitcher._upperPattern) {
        dtf = O$.getDateTimeFormatObject(daySwitcher._locale);
        daySwitcher._upperText.innerHTML = dtf.format(day, daySwitcher._upperPattern);
      }

      var dayTable = daySwitcher._getDayTable();
      if (!dayTable) {
        return;
      }

      daySwitcher._dayTable.setDay(day);
    };

    daySwitcher.previousDay = function() {
      var prevDay = O$.incDay(daySwitcher._day, -1);
      daySwitcher.setDay(prevDay);
    };

    daySwitcher.nextDay = function() {
      var nextDay = O$.incDay(daySwitcher._day, 1);
      daySwitcher.setDay(nextDay);
    };

    daySwitcher.today = function() {
      var today = new Date();
      daySwitcher.setDay(today);
    };

    if (enabled) {
      previousButton.onmousedown = function () {
        O$.setStyleMappings(previousButton, {
          pressed: stylingParams.previousButtonPressedClass
        });
        daySwitcher.previousDay();
      };

      nextButton.onmousedown = function () {
        O$.setStyleMappings(nextButton, {
          pressed: stylingParams.nextButtonPressedClass
        });
        daySwitcher.nextDay();
      };

      O$.addLoadEvent(function () {
        var dayTable = daySwitcher._getDayTable();
        if (!dayTable) {
          return;
        }

        var _onDayChange = dayTable._onDayChange;
        dayTable._onDayChange = function(day) {
          if (_onDayChange) {
            _onDayChange(day);
          }
          daySwitcher.setDay(day);
        };
      });

    }


  }
};

File: openFaces/source/org/openfaces/component/timetable/DaySwitcher.java
/*
 * OpenFaces - JSF Component Library 3.0
 * Copyright (C) 2007-2012, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */
package org.openfaces.component.timetable;

import org.openfaces.component.calendar.Calendar;
import org.openfaces.util.Components;
import org.openfaces.util.ValueBindings;

import javax.faces.context.FacesContext;

/**
 * @author Natalia Zolochevska
 */
public class DaySwitcher extends AbstractSwitcher<DayTable> {
    public static final String COMPONENT_TYPE = "org.openfaces.DaySwitcher";
    public static final String COMPONENT_FAMILY = "org.openfaces.DaySwitcher";

    private String upperDateFormat;
    private String upperPattern;

    private String upperTextStyle;
    private String upperTextClass;
    private Boolean popupCalendarEnabled;

    public DaySwitcher() {
        setRendererType("org.openfaces.DaySwitcherRenderer");
    }

    public String getUpperDateFormat() {
        return ValueBindings.get(this, "upperDateFormat", upperDateFormat);
    }

    public void setUpperDateFormat(String upperDateFormat) {
        this.upperDateFormat = upperDateFormat;
    }

    @Override
    public String getPattern() {
        String pattern = super.getPattern();
        if (pattern == null) {
            TimePeriodSwitcher timePeriodSwitcher = getTimePeriodSwitcher();
            if (timePeriodSwitcher != null)
                pattern = timePeriodSwitcher.getDayPattern();
        }
        return pattern;
    }

    public String getUpperPattern() {
        String upperPattern = ValueBindings.get(this, "upperPattern", this.upperPattern);
        if (upperPattern == null) {
            TimePeriodSwitcher timePeriodSwitcher = getTimePeriodSwitcher();
            if (timePeriodSwitcher != null)
                upperPattern = timePeriodSwitcher.getDayUpperPattern();
        }
        return upperPattern;
    }

    public void setUpperPattern(String upperPattern) {
        this.upperPattern = upperPattern;
    }

    public String getUpperTextStyle() {
        return ValueBindings.get(this, "upperTextStyle", upperTextStyle);
    }

    public void setUpperTextStyle(String upperTextStyle) {
        this.upperTextStyle = upperTextStyle;
    }

    public String getUpperTextClass() {
        return ValueBindings.get(this, "upperTextClass", upperTextClass);
    }

    public void setUpperTextClass(String upperTextClass) {
        this.upperTextClass = upperTextClass;
    }

    public String getFamily() {
        return COMPONENT_FAMILY;
    }

    @Override
    public Object saveState(FacesContext context) {
        return new Object[]{
                super.saveState(context),
                upperDateFormat,
                upperPattern,
                upperTextStyle,
                upperTextClass,
                popupCalendarEnabled
        };
    }

    @Override
    public void restoreState(FacesContext context, Object stateObj) {
        Object[] state = (Object[]) stateObj;
        int i = 0;
        super.restoreState(context, state[i++]);

        upperDateFormat = (String) state[i++];
        upperPattern = (String) state[i++];

        upperTextStyle = (String) state[i++];
        upperTextClass = (String) state[i++];

        popupCalendarEnabled = (Boolean) state[i++];
    }

    @Override
    public Timetable.ViewType getApplicableViewType() {
        return Timetable.ViewType.DAY;
    }

    public boolean isPopupCalendarEnabled() {
        return ValueBindings.get(this, "popupCalendarEnabled", popupCalendarEnabled, true);
    }

    public void setPopupCalendarEnabled(boolean popupCalendarEnabled) {
        this.popupCalendarEnabled = popupCalendarEnabled;
    }

    public Calendar getPopupCalendar() {
        Calendar calendar = Components.getOrCreateFacet(
                getFacesContext(), this, Calendar.COMPONENT_TYPE, "popupCalendar", Calendar.class);
        return calendar;
    }

}